### PR TITLE
Query component iterative.

### PR DESCRIPTION
**Pull request by claygoddard**
Add iterative support for the query component.
Additionally, add common search handler support to tie together end-to-end.

### CODE REVIEW

**Comment by claygoddard** [comment]
Just putting out an initial DO NOT MERGE PR that still needs tests alongside query grouping support. Just wanted to get any initial thoughts on structure since other components will follow suit. FWIW the query component is the most complicated one. The facet module is _much_ more straightforward.

**Comment by claygoddard** [comment]
> query grouping support

I'm actually going to push this to a separate PR as its enough to probably be looked at separately. I'll get tests added for this one and update.

**Comment by claygoddard** [comment]
Some initial tests placed in our fs-solr repo: https://github.com/cowpaths/fs-solr/pull/208 (that PR targets these lucene solr updates). This should be good for review.

File: solr/core/src/java/org/apache/solr/handler/component/ShardDoc.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/handler/component/ShardDoc.java]: Is this basically just a way to encode and decode a `ShardDoc` to something that is trivially serializable using JavaBin? If so... should this instead be a direct encode/decode here instead?

The client can then base64 encode as needed.

Not sure we need this interstitial map representation anywhere.

File: solr/core/src/java/org/apache/solr/handler/component/SearchComponent.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/handler/component/SearchComponent.java]: See my comment below. I wonder if this should just be an encode/decode that operates on opaque byte arrays instead?

Comment (claygoddard) [reply solr/core/src/java/org/apache/solr/handler/component/SearchComponent.java]: I was more concerned with compatibility if/when we ever needed to change these (same answer to your other comment). It seemed like javabin stuff may be a little more annoying there to get right. But I plan on taking a look this week at using that for serialization instead!

Comment (claygoddard) [reply solr/core/src/java/org/apache/solr/handler/component/SearchComponent.java]: The Solr `JavaBinCodec` does know how to encode and decode certain special types (like `SolrDocument`, `SolrDocumentList`, `NamedList`, etc.) which we are already getting for free at the top level as it traverses our maps and encodes the underlying values.

The main added cost with my approach is encoding the maps (with their short keys). The values will be encoded the exact same as if I did manually or let the map values be encoded.

The annoying part about using `JavaBinCodec` is on the write side managing ordering and read/write consistency. It's going to stink if we need to modify this stuff with live running requests. We'd need some form of versioning if we wanted to write raw fields one after another. On the read side, we're still going to want a top level map since there is no guarantee of search component order and existence. I'm wondering if it is actually worth the effort...

Comment (claygoddard) [reply solr/core/src/java/org/apache/solr/handler/component/SearchComponent.java]: @donjaime Pushed the latest with javabin encoding as much as possible. Still a little worried about versioning but the rest is fine. I did see maybe a 20% reduction in iterative state size (presumably due to per doc overhead which isn't typically present in most of our facet requests - although they'd have their own overhead in repeated terms and other things). LMK what you think.

**Comment by claygoddard** [comment]
I would love further eyes on this. I don't really want to forge ahead any further until we can agree on approach.

**Comment by claygoddard** [comment]
https://app.clubhouse.io/fullstory/story/164569/implement-iterative-querying-for-solr-query-component

**Review by donjaime — COMMENTED** [review]
I think things look good to me. The only outstanding thing I want to grok is what this does before we start opting in to iterative query merging. So holding off on approval until I convince myself this is a no-op prior to us adopting it.

**Comment by claygoddard** [comment]
> I think things look good to me. The only outstanding thing I want to grok is what this does before we start opting in to iterative query merging. So holding off on approval until I convince myself this is a no-op prior to us adopting it.

Yep, fair enough! Everything _should_ be a no-op when none of the new request flags are provided. I will also make another pass at this point to double check since there are some code changes in non flagged areas of the component.

File: solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java
Comment (claygoddard) [inline solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java]: I can update to only set this if `rb.isIterative()`. Potentially hold onto less memory in the non iterative case.

File: solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java
Comment (claygoddard) [inline solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java]: I'm going to update to always overwrite the component state here.

**Comment by chatman** [comment]
Here's our understanding of the design and the implementation, mainly gathered from looking at the unit tests in the fs-solr branch. Please let us know if we're missing something.

Approach: The client would break up a query into sub-queries with disjoint ranges, issue them one after the other, while passing on the previous step's results ("iterative_state") to the next sub-query every time. At every step, the iterative_state passed in would be merged with current results and returned.

In this approach, I sense that in the last iteration, the request will send in all results received until then (iterative_state) to Solr and Solr will prepare a response by merging the iterative_state with the last step's results and return the final results. In that particular iteration (the last one), the memory consumed will likely be not less than as much as would have been consumed in the full query (that this approach aims to replace) since it will generate the same result set after the merging. In addition to that, every step will require additional resources on the Solr nodes for encoding/decoding the iterative_step for passing it back to the client and reading it back from the request for next iteration. Does this sound like a fair assessment of the situation?

If the above understanding is accurate, then Noble and I had an idea to optimize this. The idea is to send one query to every shard (that covers the full range of results intended for that shard), and that query can be intercepted by a custom request handler/query component on the Solr node, and that RH/QC rewrites that query to other subqueries (breaks them into further sub-ranges within that shard) and issues them iteratively (as is achieved in this approach), and returns to client the fully merged result (for that shard). This results in lowering the overhead in passing iterative_state back and forth, since the RH/QC would merge all of that in the scope of the same request.

**Comment by claygoddard** [comment]
> Here's our understanding of the design and the implementation, mainly gathered from looking at the unit tests in the fs-solr branch. Please let us know if we're missing something.
> 
> Approach: The client would break up a query into sub-queries with disjoint ranges, issue them one after the other, while passing on the previous step's results ("iterative_state") to the next sub-query every time. At every step, the iterative_state passed in would be merged with current results and returned.

Yes, this is accurate. Of note, when we say sub-query, the iterative state only lives on QA - just as a clarification. But I think this assumes that correctly.
 
> In this approach, I sense that in the last iteration, the request will send in all results received until then (iterative_state) to Solr and Solr will prepare a response by merging the iterative_state with the last step's results and return the final results.

This will be happening each step of the way right now as it was simpler to deal with _one_ iterative state. But we could optimize to allow any number of iterative states to be passed at a later step (like only the final step).

> In that particular iteration (the last one), the memory consumed will likely be not less than as much as would have been consumed in the full query (that this approach aims to replace) since it will generate the same result set after the merging. In addition to that, every step will require additional resources on the Solr nodes for encoding/decoding the iterative_step for passing it back to the client and reading it back from the request for next iteration. Does this sound like a fair assessment of the situation?

Agreed, the memory used by the end will be roughly the same (and honestly mostly throughout given cardinality of fields, doc limits, etc.). All in all, this is a fair assessment.

> If the above understanding is accurate, then Noble and I had an idea to optimize this. The idea is to send one query to every shard (that covers the full range of results intended for that shard), and that query can be intercepted by a custom request handler/query component on the Solr node, and that RH/QC rewrites that query to other subqueries (breaks them into further sub-ranges within that shard) and issues them iteratively (as is achieved in this approach), and returns to client the fully merged result (for that shard).

Yeah, @donjaime had suggested a similar approach as this at one point too.

There are few things that makes this difficult and a few downsides.

1. Solr would need to know exactly where to do the slicing and with which field over what ranges. The latter could be a range facet like configuration supplied in the request. The former would need some kind of notation to indicate which groups of expressions in entirety would be sliced. I guess possibly a query parser that injects a sliced expression could work here - actually the query parser itself could just intersect with a precomputed docset for the current contextual slice. Still it gets more complicated for Solr to have to understand the slicing.
2. Timeouts / time allowed. Are there any limitations to how long Solr will wait on a shard? Any other side effects to way longer lived per shard requests? We can't as easily enforce time allowed on the client side here as it seems less predictable.
3. Iterative / partial results. This approach would eliminate the ability to get initial iterative / partial / sampled results (unless we streamed something back along the way which seems complicated in this setup) which feels like it is becoming a key part of our LTA / expensive query solution for quick feedback and/or optional sample only queries for quick iteration.
4. Pick up where you left off. If this fails at any point, you can't easily start where you left off without some level of persistence in Solr.

But I totally agree there are upsides.

1. Remove overhead of repeated network requests to all shards.
2. Client needs no state management to perform these iterative requests.
3. All search components *just work* when aggregating / merging shard response (there are few downsides to my approach where we don't always have the full view of data at each step).

> This results in lowering the overhead in passing iterative_state back and forth, since the RH/QC would merge all of that in the scope of the same request.

Just a note here, this overhead is _only_ seen by the Solr client and the QA node. The distributed shard requests do not see this overhead. But the distributed shard requests themselves are overhead - having to make network requests to _all_ shards for each iteration.

I feel like with cons 3. & 4. it gets hard to see this approach (even with the upsides) as being a viable option. Thoughts?

**Comment by claygoddard** [comment]
(also going to throw these comments in an appendix of my design once we come to a resolution)

**Review by donjaime — APPROVED** [review]
LGTM from me given that all the logic is now behind either an `isIterative()` check or behind an `json.containsKey("iterative_state")` check.

**Comment by claygoddard** [comment]
Going to merge to have for further testing as we continue the conversation. Facet module changes to come soon (they are simpler!).

**Pull request merged**
